datafusion/src/physical_plan/expressions/bitmap_distinct.rs

```diff
@@ -0,0 +1,233 @@
+                        .iter()
+                        .map(|&x| x as u32)
+                        .collect::<Vec<u32>>(),
+                );
+            }
+            DataType::Int16 => {
+                let array = value.as_any().downcast_ref::<Int16Array>().unwrap();
+                self.bitmap.add_many(
+                    &array
+                        .values()
+                        .iter()
+                        .map(|&x| x as u32)
+                        .collect::<Vec<u32>>(),
+                );
+            }
+            DataType::Int32 => {
+                let array = value.as_any().downcast_ref::<Int32Array>().unwrap();
+                self.bitmap.add_many(
+                    &array
+                        .values()
+                        .iter()
+                        .map(|&x| x as u32)
+                        .collect::<Vec<u32>>(),
+                );
+            }
+            DataType::UInt8 => {
+                let array = value.as_any().downcast_ref::<UInt8Array>().unwrap();
+                self.bitmap.add_many(
+                    &array
+                        .values()
+                        .iter()
+                        .map(|&x| x as u32)
+                        .collect::<Vec<u32>>(),
+                );
+            }
+            DataType::UInt16 => {
+                let array = value.as_any().downcast_ref::<UInt16Array>().unwrap();
+                self.bitmap.add_many(
+                    &array
+                        .values()
+                        .iter()
+                        .map(|&x| x as u32)
+                        .collect::<Vec<u32>>(),
+                );
+            }
+            DataType::UInt32 => {
+                let array = value.as_any().downcast_ref::<UInt32Array>().unwrap();
+                self.bitmap.add_many(array.values());
+            }
+            e => {
+                return Err(DataFusionError::Internal(format!(
+                    "BITMAP_COUNT_DISTINCT is not expected to receive the type {:?}",
+                    e
+                )));
+            }
+        }
+        Ok(())
+    }
+
+    fn merge_batch(&mut self, states: &[ArrayRef]) -> Result<()> {
+        assert_eq!(1, states.len(), "expect only 1 element in the states");
+        let binary_array = states[0].as_any().downcast_ref::<BinaryArray>().unwrap();
+        let bitmaps = binary_array
+            .iter()
+            .map(|x| {
+                Bitmap::deserialize(x.expect(
+                    "Impossibly got empty binary array from states in bitmap_distinct",
+                ))
+            })
+            .collect::<Vec<Bitmap>>();
+        let bitmaps = &bitmaps.iter().map(|x| x.borrow()).collect::<Vec<&Bitmap>>()[..];
+        //Do not use self.bitmap = xxx, because '=' has been wrote for 'BitAnd' !.
+        self.bitmap.bitor_assign(Bitmap::fast_or(bitmaps));
+        Ok(())
+    }
+
+    fn evaluate(&self) -> Result<ScalarValue> {
```

Comment:
I assume this generally only gets called once per query? Only asking because `cardinality` is quite an expensive call.

Reply:
Sorry , is there something i miss ? Are there more efficient API, to get the number of integers contained in the bitmap

Reply:
No this is the only way. I was just checking it is only called once for the final result.